File: packages/client/src/schemas/storage.ts
import { z } from 'zod';

/**
 * Exact source-map storage weight, summed from the `size` columns of the
 * in-DB chunk table and the on-disk source_file CAS.
 */
export const sourceMapStorageSchema = z.object({
  chunk_bytes: z.number(),
  source_file_bytes: z.number(),
  total_bytes: z.number(),
  file_count: z.number(),
});

/**
 * Instance-wide storage summary: row counts per data category, whole-DB size
 * (best-effort, backend-reported), and exact source-map weight.
 */
export const storageSummarySchema = z.object({
  total_db_size_bytes: z.number(),
  events_count: z.number(),
  transactions_count: z.number(),
  spans_count: z.number(),
  logs_count: z.number(),
  source_maps: sourceMapStorageSchema,
});

/**
 * Per-project storage breakdown (one row per project, including empty ones).
 * `estimated_bytes` is the summed JSON payload length the project owns.
 */
export const projectStorageSchema = z.object({
  project_id: z.number(),
  project_name: z.string(),
  events_count: z.number(),
  transactions_count: z.number(),
  spans_count: z.number(),
  logs_count: z.number(),
  source_maps_count: z.number(),
  estimated_bytes: z.number(),
});

/**
 * Rows affected by a cleanup — used symmetrically for the dry-run preview
 * ("what would be removed") and the executed result ("what was removed").
 */
export const cleanupCountsSchema = z.object({
  events: z.number(),
  transactions: z.number(),
  spans: z.number(),
  logs: z.number(),
  issues_removed: z.number(),
});

/**
 * Outcome of a source-map garbage collection: orphaned files removed from the
 * DB and unlinked from disk, plus the exact bytes freed.
 */
export const sourceMapGcResultSchema = z.object({
  files_removed: z.number(),
  bytes_freed: z.number(),
});

/**
 * Request body for a cleanup (preview or execute): remove data older than
 * `older_than_days`, optionally scoped to one project (omit for all) and to
 * specific data categories. The minimum of 1 mirrors the server contract — a
 * smaller window would wipe the whole dataset.
 *
 * The `include_*` flags select which data categories the cleanup acts on. Each
 * is optional and defaults to `true` server-side, so omitting all of them keeps
 * the "delete everything older than the cutoff" behaviour. `include_transactions`
 * also governs the cascaded spans, and `include_events` governs the emptied
 * issues that get removed alongside their events.
 */
export const cleanupOptionsSchema = z.object({
  older_than_days: z.number().int().min(1),
  project_id: z.number().int().optional(),
  include_events: z.boolean().optional(),
  include_transactions: z.boolean().optional(),
  include_logs: z.boolean().optional(),
});

File: apps/server/src/models/storage.rs
use serde::{Deserialize, Serialize};

/// Request body for a cleanup (preview or execute): remove data older than
/// `older_than_days`, optionally scoped to a single project (omit for all).
///
/// `older_than_days` must be at least 1 — the service rejects smaller values,
/// since a window of 0 (or negative) would move the cutoff to now/the future and
/// purge the entire dataset.
#[derive(Debug, Deserialize)]
#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
pub struct CleanupRequest {
    #[cfg_attr(feature = "openapi", schema(minimum = 1))]
    pub older_than_days: i64,
    #[serde(default)]
    pub project_id: Option<i32>,
}

/// Instance-wide storage summary: what Rustrak is holding right now.
///
/// Row counts are exact (indexed `COUNT(*)`). `total_db_size_bytes` is a
/// best-effort figure from the backend (`pg_database_size` / SQLite page count),
/// while `source_maps` is exact (summed `size` columns).
#[derive(Debug, Default, Serialize, PartialEq)]
#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
pub struct StorageSummary {
    /// Whole-database size in bytes (best-effort, backend-reported).
    pub total_db_size_bytes: i64,
    pub events_count: i64,
    pub transactions_count: i64,
    pub spans_count: i64,
    pub logs_count: i64,
    /// Exact source-map weight + file count.
    pub source_maps: SourceMapStorage,
}

/// Rows affected by a cleanup, used symmetrically for the dry-run preview ("what
/// would be removed") and the executed result ("what was removed"). `spans` are
/// counted directly even though they cascade from `transactions`, so the UI can
/// show the true row impact. `issues_removed` is the number of issues that would
/// be left with zero events and therefore deleted.
#[derive(Debug, Default, Serialize, PartialEq)]
#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
pub struct CleanupCounts {
    pub events: i64,
    pub transactions: i64,
    pub spans: i64,
    pub logs: i64,
    pub issues_removed: i64,
}

/// Outcome of a source-map garbage collection: orphaned `source_file` rows
/// (no `source_file_metadata` referencing them) removed from the DB and unlinked
/// from disk. `bytes_freed` is the exact summed `size` of the removed files.
#[derive(Debug, Default, Serialize, PartialEq)]
#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
pub struct SourceMapGcResult {
    pub files_removed: i64,
    pub bytes_freed: i64,
}

/// Per-project storage breakdown, one row per project (including empty ones).
///
/// Counts are exact. `estimated_bytes` is the summed length of the JSON payloads
/// this project owns across events/transactions/spans — a real per-project weight,
/// not an apportioned guess.
#[derive(Debug, Default, Serialize, PartialEq)]
#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
pub struct ProjectStorage {
    pub project_id: i32,
    pub project_name: String,
    pub events_count: i64,
    pub transactions_count: i64,
    pub spans_count: i64,
    pub logs_count: i64,
    pub source_maps_count: i64,
    pub estimated_bytes: i64,
}

/// Exact source-map storage weight, summed from the `size` columns.
///
/// Source maps live in two places: `chunk` holds the raw upload bytes as in-DB
/// BYTEA, `source_file` holds the assembled artifact on the filesystem CAS. Both
/// carry a `size` column, so the weight is an exact SUM — no filesystem walk.
#[derive(Debug, Default, Serialize, PartialEq)]
#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
pub struct SourceMapStorage {
    /// Bytes held in the in-DB `chunk` table (raw upload chunks).
    pub chunk_bytes: i64,
    /// Bytes held on disk via `source_file` (assembled artifacts).
    pub source_file_bytes: i64,
    /// `chunk_bytes + source_file_bytes`.
    pub total_bytes: i64,
    /// Number of `source_file` rows.
    pub file_count: i64,
}
